### PR TITLE
Implement `KJ_CO_MAGIC BEGIN_DEFERRED_PROXYING`

### PR DESCRIPTION
**Pull request by harrishancock**
Depends on: https://github.com/capnproto/capnproto/pull/1741

This commit implements a coroutine adapter for a new Promise-derived class, `DeferredProxyPromise<T>`. The purpose of the new coroutine adapter is to allow such coroutines to use the expression `co_yield BEGIN_DEFERRED_PROXYING` to indicate that the deferred proxying stage of their operation has begun. The `DeferredProxyPromise<T>` coroutine return type is itself derived from `kj::Promise<T>`, and can be used like one.

This allows the following transformations:

---
```cpp
kj::Promise<DeferredProxy<void>> foo() {
  // Enter deferred proxying with an immediately-ready `proxyTask`.
  return newNoopDeferredProxy();
}
```
becomes
```cpp
DeferredProxyPromise<void> foo() {
  co_return;
}
```
---
```cpp
kj::Promise<DeferredProxy<void>> foo() {
  // Enter deferred proxying with a promise that should be awaited as `proxyTask`.
  return DeferredProxy<void> { kj::mv(promise) };
}
```
becomes
```cpp
DeferredProxyPromise<void> foo() {
  co_yield BEGIN_DEFERRED_PROXYING;
  co_return co_await promise;
}
```
---
```cpp
kj::Promise<DeferredProxy<void>> foo() {
  // Wait for `promise` to be done, then enter deferred proxying with an immediately-ready `proxyTask`.
  return addNoopDeferredProxy(promise);
}
```
becomes
```cpp
DeferredProxyPromise<void> foo() {
  co_return co_await promise;
}
```

### CODE REVIEW

**Review by mikea — COMMENTED** [review]
I can't comment much on the actual implementation of coroutine handler, but the API introduced looks very clear and straightforward.

File: src/workerd/api/web-socket.c++
Comment (jasnell) [inline src/workerd/api/web-socket.c++:336]: Definitely like this syntax but it would likely be good to add some comments around it to help new folks understand it.

File: src/workerd/api/util.h
Comment (jasnell) [inline src/workerd/api/util.h:156]: Non-blocking nit: Any chance this could be moved out into its own header, and possibly moved into `src/workerd/util`? Just want to keep `util.h` itself from becoming too much Kitchen Sink.

File: src/workerd/api/web-socket.c++
Comment (harrishancock) [reply src/workerd/api/web-socket.c++:339]: I added a comment.

File: src/workerd/api/util.h
Comment (harrishancock) [reply src/workerd/api/util.h]: I agree, done in the last commit.

File: src/workerd/api/global-scope.c++
Comment (jasnell) [inline src/workerd/api/global-scope.c++:21]: Non-blocking nit: is the new util import still needed here?

**Comment by harrishancock** [comment]
One thing I'm uncertain about is the need for a new coroutine return type -- `DeferredProxyPromise<T>`. I think it should be possible to simply specialize the coroutine adapter for `kj::Promise<DeferredProxy<T>>`, which would decouple the implementation (custom coroutine) from the return type (`kj::Promise<DeferredProxy<T>>`).

The downside to specializing like that means that if someone really does want to write `co_return DeferredProxy<T>` for some reason, they can't -- they can only write coroutines that support the weird new `co_yield` thing.

**Comment by harrishancock** [comment]
Updated to specialize `kj::Promise<DeferredProxy<T>>`'s coroutine adapter instead of creating a new coroutine return type.

I also changed the implementation of the new coroutine adapter to be a PromiseNode, which allows us to avoid all extra heap allocations.

**Comment by kentonv** [comment]
Neat!

Do we really want people to see `co_yield` here? This could perhaps be distracting for people who have used `co_yield` for its intended purpose -- it implies that the coroutine actually "yields" here waiting for something to call it back asking for the next value, but that's not actually what is happening. We're actually hijacking `co_yield` as an entrypoint to the coroutine adapter in order to do something completely different.

To that end I'd be inclined to define a macro:

```
#define WORKERD_BEGIN_DEFERRED_PROXYING co_yield BEGIN_DEFERRED_PROXYING
```

This leaves people seeing just a magic macro at the usage site, which only tells them that there is magic, without misleading them into thinking it yields.

**Comment by harrishancock** [comment]
In this particular case, `co_yield` makes sense to my brain, but I like the idea of having `co_magic ...` as the one entrypoint to invoke magic coroutine stuff. Otherwise I fear we'll have a proliferation of different `co_blah` macros as we think up new things.

Defining `co_magic` literally as a synonym for `co_yield` worries me a little -- will the error messages be helpful, and will people implementing generators ever get confused? I thought about suggesting `#define co_magic co_yield ::kj::_::MAGIC *`, where `decltype(kj::_::MAGIC)::operator*(T)` returns a `MagicBox<T>`. That would make all magic calls to `yield_value()` take a `MagicBox<T>` instead of a `T` ... but I'm not sure that actually solves either problem, so maybe never mind.

**Comment by harrishancock** [comment]
1. Rebased on main branch.
2. Picked up changes in this capnproto PR: https://github.com/capnproto/capnproto/pull/1754
3. Use `co_magic` instead of `co_yield`.
4. DeferredProxyCoroutine passes its Args... through to the kj::Promise coroutine, to benefit from the inner coroutine's validation that no parameters are a WaitScope.
5. Added a `yield_value()` overload which forwards any unrecognized value types through to the inner coroutine's `yield_value()` function, if it has one (it is SFINAE-friendly). This means any magic that regular kj::Promise coroutines gain will be automatically available in DeferredProxyCoroutine promises, too.

File: src/workerd/api/deferred-proxy-test.c++
Comment (kentonv) [inline src/workerd/api/deferred-proxy-test.c++:1]: Does this test cover cancellation, both before and after starting deferred proxying?

**Comment by harrishancock** [comment]
1. Rebased on main branch.
2. Switched to this, more targeted capnproto PR: https://github.com/capnproto/capnproto/pull/1759
3. Use `KJ_CO_MAGIC` instead of `co_magic`.

File: src/workerd/api/deferred-proxy-test.c++
Comment (harrishancock) [reply src/workerd/api/deferred-proxy-test.c++:1]: It did not; added in ca266c205623186535fff68a8c1edb4c2348f07f.

**Comment by harrishancock** [comment]
Rebased on `main` branch.

**Comment by harrishancock** [comment]
Updated capnproto submodule to latest `v2` branch.

File: src/workerd/api/deferred-proxy-test.c++
Comment (jasnell) [inline src/workerd/api/deferred-proxy-test.c++:56]: Looking at this, if we're going to define a kj macro for this, we may as well define it for the entire line... e.g.

e.g.
```
KJ_CO_BEGIN_DEFERRED_PROXYING;
```

Or do we imagine that `KJ_CO_MAGIC` will be used for other cases?

File: src/workerd/api/deferred-proxy.h
Comment (kentonv) [inline src/workerd/api/deferred-proxy.h:221]: I'm struggling to understand this. How does this object get destroyed if destroy() is a no-op?

Have you run your test under a memory leak checker?

Comment (kentonv) [reply src/workerd/api/deferred-proxy.h:221]: I guess the answer is that `inner` holds onto the outer coroutine handle and so inner's `destroy()` actually calls `handle.destroy()` which destroys the outer coroutine... this is very mind-bendy.

Comment (kentonv) [reply src/workerd/api/deferred-proxy.h:221]: What happens if `destroy()` ends up being called on the outer PromiseNode _after_ the inner one is destroyed? Then I guess this method is called on an already-freed object (probably segfaulting on the vtable). Can we definitively say this can't happen in practice?

File: src/workerd/api/deferred-proxy-test.c++
Comment (harrishancock) [reply src/workerd/api/deferred-proxy-test.c++:56]: Yes, we do: https://github.com/capnproto/capnproto/pull/1739

File: src/workerd/api/deferred-proxy.h
Comment (harrishancock) [reply src/workerd/api/deferred-proxy.h]: > Have you run your test under a memory leak checker?

Valgrind reports no leaks.

> this is very mind-bendy

Yep.

> What happens if destroy() ends up being called on the outer PromiseNode after the inner one is destroyed? Then I guess this method is called on an already-freed object (probably segfaulting on the vtable). Can we definitively say this can't happen in practice?

That's a great point -- incorrect destruction order would cause a segfault, so the second paragraph of this comment that I wrote is BS.

I think we _can_ definitively say that this won't happen in practice, because if kj-async failed to destroy PromiseNodes immediately after calling `get()`, then we'd have other problems. For example, some `.then()` continuations may rely on `kj::defer()`s running inside of attachments to their dependency promises before the continuations are invoked -- if those dependencies were not destroyed, then that guarantee would no longer hold. If you grep through async.c++, you'll see that every call to `get()` is immediately succeeded by a nullification of the node whose `get()` was called. (Except ironically, the one that backs `co_await`. That one is safe because the PromiseNode is in a temporary value that is destroyed after the `co_await expression is evaluated, though perhaps we should make that explicit.)

It'd be nice to have a stronger guarantee, though, so I pushed a commit which adds a test which forces a `get()` without immediate dependency destruction, and adds a mechanism to destroy the outer PromiseNode in its own `get()` via its self-pointer. Let me know what you think.

Aside: it has occurred to me that perhaps `get()` should somehow imply `destroy()`, and that `onReady()` should somehow imply `setSelfPointer()`...

Comment (kentonv) [reply src/workerd/api/deferred-proxy.h]: Hmm, I think I'm happy with your argument for why this was already a safe assumption and it's not necessary to add complexity to guarantee it. I'll approve this and leave it up to you if you want to revert that last commit (maybe keeping the test?).

Meanwhile, though, I do like the idea of `get()` implying `destroy()`. This might be a nice optimization for KJ promises across the board, avoiding a redundant virtual call. (But that would be a separate PR...)

**Review by kentonv — APPROVED** [review]

File: src/workerd/api/deferred-proxy.h
Comment (harrishancock) [reply src/workerd/api/deferred-proxy.h]: The Mac build seemed to fail spuriously, so I force-pushed a rebase on `main`. The HEAD commit of this PR is still identical to the previous 82d2aff.

Comment (harrishancock) [reply src/workerd/api/deferred-proxy.h]: > Hmm, I think I'm happy with your argument for why this was already a safe assumption and it's not necessary to add complexity to guarantee it. 

I feel the same ... but I think I'd like to keep the safeguard, at least until `get()` can imply `destroy()`. The reason is that I have personally introduced exactly such incorrect destruction order bugs in kj-async while iterating on unrelated async PRs in the now-distant past, and it'd be nice if they were very noisy failures. The new test case I introduced didn't actually fail until I ran it under Valgrind, underscoring how insidious memory access problems are. :/

> I'll approve this and leave it up to you if you want to revert that last commit (maybe keeping the test?).

We couldn't actually keep the test if we remove the safeguard, since the test intentionally misuses the private `kj::_` API to trigger the problem.

**Pull request merged**
